File: frontend/src/testing/commands/BaseCommand.ts
/**
 * BaseCommand - Common interfaces and utilities for all test commands
 */

export interface TestContext {
  page?: any; // Playwright page
  browser?: any; // Playwright browser
  baseUrl: string;
  timeout: number;
  retries: number;
  failFast: boolean;
}

export interface TestResult {
  name: string;
  success: boolean;
  duration: number;
  errors: string[];
  warnings: string[];
  artifacts: string[];
  metadata: Record<string, any>;
  // Fail-fast specific fields
  shouldStopPipeline?: boolean;
  criticalError?: boolean;
  issueId?: string;
  remediation?: string;
}

export interface CommandInput {
  timeout?: number;
  retries?: number;
  failFast?: boolean;
}

// Global timeout configurations
export const GLOBAL_TIMEOUTS = {
  // Page operations
  pageLoad: 30000,        // 30s for page load
  navigation: 15000,      // 15s for navigation
  elementWait: 10000,     // 10s for element waits
  click: 5000,            // 5s for clicks
  
  // Map operations
  mapLoad: 20000,         // 20s for map initialization
  mapReady: 15000,        // 15s for map to be ready
  layerLoad: 10000,       // 10s for layer loading
  layerRender: 8000,      // 8s for layer rendering
  
  // Validation operations
  validation: 10000,      // 10s for validation checks
  performance: 5000,      // 5s for performance measurements
  interaction: 3000,      // 3s for interactions
  
  // File operations
  screenshot: 8000,       // 8s for screenshot capture
  fileWrite: 5000,        // 5s for file operations
  
  // Network operations
  apiCall: 10000,         // 10s for API calls
  networkIdle: 5000,      // 5s for network idle
  
  // Backend operations
  database: 15000,        // 15s for database operations
  serviceStart: 30000,    // 30s for service startup
  healthCheck: 10000      // 10s for health checks
};

// Helper function to add timeout to any promise
export function withTimeout<T>(
  promise: Promise<T>, 
  timeoutMs: number, 
  errorMessage: string
): Promise<T> {
  return Promise.race([
    promise,
    new Promise<T>((_, reject) => 
      setTimeout(() => reject(new Error(`${errorMessage} (timeout after ${timeoutMs}ms)`)), timeoutMs)
    )
  ]);
}

// Helper function to retry operations with exponential backoff
export async function withRetry<T>(
  operation: () => Promise<T>,
  maxRetries: number = 3,
  baseDelay: number = 1000
): Promise<T> {
  let lastError: Error;
  
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await operation();
    } catch (error) {
      lastError = error as Error;
      
      if (attempt === maxRetries) {
        throw lastError;
      }
      
      const delay = baseDelay * Math.pow(2, attempt);
      await new Promise(resolve => setTimeout(resolve, delay));
    }
  }
  
  throw lastError!;
}

// Base command class that all commands should extend
export abstract class BaseCommand {
  abstract name: string;
  protected input: CommandInput;
  protected startTime: number = 0;

  constructor(input: CommandInput = {}) {
    this.input = {
      timeout: GLOBAL_TIMEOUTS.validation,
      retries: 1,
      failFast: true,
      ...input
    };
  }

  abstract run(ctx: TestContext): Promise<TestResult>;

  protected createResult(
    success: boolean, 
    errors: string[] = [], 
    warnings: string[] = [], 
    artifacts: string[] = [], 
    metadata: Record<string, any> = {},
    shouldStopPipeline: boolean = false,
    criticalError: boolean = false,
    issueId?: string,
    remediation?: string
  ): TestResult {
    return {
      name: this.name,
      success,
      duration: Date.now() - this.startTime,
      errors,
      warnings,
      artifacts,
      metadata,
      shouldStopPipeline: shouldStopPipeline || !success,
      criticalError: criticalError || !success,
      issueId: issueId || (errors.length > 0 ? `ISSUE_${Date.now()}_${Math.random().toString(36).substr(2, 9)}` : undefined),
      remediation: remediation || this.generateRemediation(errors)
    };
  }

  protected generateRemediation(errors: string[]): string {
    if (errors.length === 0) return '';
    
    const error = errors[0].toLowerCase();
    
    if (error.includes('timeout')) {
      return 'Increase timeout or check network connectivity';
    } else if (error.includes('not found') || error.includes('missing')) {
      return 'Verify element selectors and page structure';
    } else if (error.includes('permission') || error.includes('access')) {
      return 'Check authentication and authorization settings';
    } else if (error.includes('network') || error.includes('connection')) {
      return 'Verify network connectivity and service availability';
    } else if (error.includes('performance') || error.includes('slow')) {
      return 'Optimize performance or increase performance thresholds';
    } else if (error.includes('security') || error.includes('https')) {
      return 'Fix security configuration and SSL certificates';
    } else {
      return 'Review error details and check system logs';
    }
  }

  protected async executeWithTimeout<T>(
    operation: () => Promise<T>,
    timeoutMs: number = this.input.timeout!,
    errorMessage: string = `${this.name} operation timeout`
  ): Promise<T> {
    return withTimeout(operation(), timeoutMs, errorMessage);
  }

  protected async executeWithRetry<T>(
    operation: () => Promise<T>,
    maxRetries: number = this.input.retries!,
    baseDelay: number = 1000
  ): Promise<T> {
    return withRetry(operation, maxRetries, baseDelay);
  }
}

File: backend/tests/commands/BaseCommand.py
"""
BaseCommand - Common interfaces and utilities for all backend test commands
"""

import time
import asyncio
from abc import ABC, abstractmethod
from typing import Dict, List, Any, Optional, Union
from dataclasses import dataclass
from enum import Enum

class TestStatus(Enum):
    SUCCESS = "success"
    FAILURE = "failure"
    SKIPPED = "skipped"
    ERROR = "error"

@dataclass
class TestContext:
    """Context for test execution"""
    base_url: str
    timeout: int
    retries: int
    fail_fast: bool
    environment: str = "test"
    database_url: Optional[str] = None
    api_key: Optional[str] = None

@dataclass
class TestResult:
    """Result of test execution"""
    name: str
    status: TestStatus
    duration: float
    errors: List[str]
    warnings: List[str]
    artifacts: List[str]
    metadata: Dict[str, Any]

# Global timeout configurations
GLOBAL_TIMEOUTS = {
    # API operations
    "api_call": 10,           # 10s for API calls
    "api_response": 15,        # 15s for API responses
    "health_check": 10,        # 10s for health checks
    
    # Database operations
    "database_query": 15,      # 15s for database queries
    "database_connection": 10, # 10s for database connections
    "database_migration": 30,  # 30s for database migrations
    
    # Service operations
    "service_start": 30,       # 30s for service startup
    "service_stop": 10,        # 10s for service shutdown
    "service_health": 10,      # 10s for service health checks
    
    # File operations
    "file_read": 5,            # 5s for file reads
    "file_write": 5,           # 5s for file writes
    "file_upload": 15,         # 15s for file uploads
    
    # Validation operations
    "validation": 10,          # 10s for validation checks
    "data_processing": 20,     # 20s for data processing
    "model_inference": 30,     # 30s for ML model inference
    
    # Network operations
    "network_request": 15,     # 15s for network requests
    "network_timeout": 30,     # 30s for network timeouts
    "connection_pool": 5,      # 5s for connection pool operations
}

class TimeoutError(Exception):
    """Raised when an operation times out"""
    pass

class RetryError(Exception):
    """Raised when all retry attempts are exhausted"""
    pass

def with_timeout(operation, timeout_seconds: int, error_message: str = "Operation timeout"):
    """Add timeout to any operation"""
    async def timeout_wrapper():
        try:
            if asyncio.iscoroutinefunction(operation):
                return await asyncio.wait_for(operation(), timeout=timeout_seconds)
            else:
                return await asyncio.wait_for(
                    asyncio.get_event_loop().run_in_executor(None, operation), 
                    timeout=timeout_seconds
                )
        except asyncio.TimeoutError:
            raise TimeoutError(f"{error_message} (timeout after {timeout_seconds}s)")
    
    return timeout_wrapper()

def with_retry(operation, max_retries: int = 3, base_delay: float = 1.0):
    """Retry operation with exponential backoff"""
    last_error = None
    
    for attempt in range(max_retries + 1):
        try:
            if asyncio.iscoroutinefunction(operation):
                return asyncio.run(operation())
            else:
                return operation()
        except Exception as error:
            last_error = error
            
            if attempt == max_retries:
                raise RetryError(f"Operation failed after {max_retries} retries: {last_error}")
            
            delay = base_delay * (2 ** attempt)
            time.sleep(delay)
    
    raise last_error

class BaseCommand(ABC):
    """Base command class that all commands should extend"""
    
    def __init__(self, input_data: Dict[str, Any] = None):
        self.input = input_data or {}
        self.start_time = 0
        self.name = self.__class__.__name__
    
    @abstractmethod
    async def run(self, ctx: TestContext) -> TestResult:
        """Execute the command"""
        pass
    
    def create_result(
        self, 
        status: TestStatus, 
        errors: List[str] = None, 
        warnings: List[str] = None, 
        artifacts: List[str] = None, 
        metadata: Dict[str, Any] = None
    ) -> TestResult:
        """Create a test result"""
        return TestResult(
            name=self.name,
            status=status,
            duration=time.time() - self.start_time,
            errors=errors or [],
            warnings=warnings or [],
            artifacts=artifacts or [],
            metadata=metadata or {}
        )
    
    async def execute_with_timeout(
        self, 
        operation, 
        timeout_seconds: int = None, 
        error_message: str = None
    ):
        """Execute operation with timeout"""
        timeout = timeout_seconds or GLOBAL_TIMEOUTS.get("validation", 10)
        error_msg = error_message or f"{self.name} operation timeout"
        
        return await with_timeout(operation, timeout, error_msg)
    
    async def execute_with_retry(
        self, 
        operation, 
        max_retries: int = None, 
        base_delay: float = 1.0
    ):
        """Execute operation with retry"""
        retries = max_retries or self.input.get("retries", 1)
        return await with_retry(operation, retries, base_delay)
